File: day-7/part-2.py
FIRST = len('Step ')
SECOND = len('Step C must be finished before step ')

class Node:
 def __init__(self, key):
   self.incoming = []
   self.outgoing = []
   self.key = key

 def __repr__(self):
   return self.key

 def __str__(self):
   return self.key

def getOrAdd(graph, key):
  if key not in graph:
    graph[key] = Node(key)
  return graph[key]

class Worker:
  def __init__(self):
    self.time = 0
    self.end = 0
    self.work = None

  def tick(self):
    self.time += 1
    return self.time >= self.end

  def set_work(self, node, end=0):
    self.work = node
    self.time = 0
    self.end = end

with open('input') as f:
  lines = f.readlines()
  pairs = map(lambda line: (line[FIRST], line[SECOND]), lines)
  graph = {}
  for (a, b) in pairs:
    nodeA = getOrAdd(graph, a)
    nodeB = getOrAdd(graph, b)
    nodeA.outgoing.append(b)
    nodeB.incoming.append(a)
  
  queue = filter(lambda node: len(node.incoming) == 0, graph.values())
  for node in queue:
    del graph[node.key]
  
  workers = [Worker() for _ in range(5)]

  def process_node(current):
    for out in current.outgoing:
      if out not in graph:
        continue
      other = graph[out]
      other.incoming.remove(current.key)
      if len(other.incoming) == 0:
        queue.append(other)
        del graph[out]
 
  counter = 0
  while True: 
    available = filter(lambda worker: worker.tick(), workers)
    for a in available:
      if a.work:
        process_node(a.work)
        # Reset worker
        a.set_work(None)

    queue.sort(key=lambda node: node.key)
    for worker in available:
      if len(queue) == 0:
        break
      current = queue.pop(0)
      if current:
        worker.set_work(current, 61 + ord(current.key) - ord('A'))
    if len(filter(lambda worker: worker.work, workers)) == 0:
      break
    else:
      counter += 1
  print str(counter)

File: day-7/part-2.js
const { readLines } = require('../util.js');

const LINE = 'Step X must be finished before step ';

function getNode(map, id) {
  let node = map.get(id);
  if (!node) {
    map.set(id, node = {id, requirements:[], nodes: []});
  }
  return node;
}

function tickWorkers(workers) {
  return workers.filter(worker => {
    return ++worker.time >= worker.end;
  })
}

readLines('input')
  .then(lines => {
    return lines.map(line => {
      let requirement = line.slice('Step '.length)[0];
      let node = line.slice(LINE.length)[0];
      return {requirement, node};
    }).reduce((graph, edge) => {
      let r = getNode(graph, edge.requirement);
      let n = getNode(graph, edge.node);
      n.requirements.push(r.id);
      r.nodes.push(n.id);
      return graph;
    }, new Map());
  }).then(graph => {
    let queue = [...graph.entries()]
                  .filter(([id, node]) => !node.requirements.length)
                  .map(([id, node]) => node);
    queue.forEach(node => graph.delete(node.id));
    let sleep = new Map([..."abcdefghijklmnopqrstuvwxyz"]
                  .map((c, i) => [c.toUpperCase(), i + 61]));
    let workers = Array(5).fill(0).map(_ => ({time: 0, end: 0}))
    let counter = 0;
    while (true) {
      let available = tickWorkers(workers);
      available.filter(a => a.work).forEach(a => {
        let node = a.work;
        node.nodes.forEach(other => { 
          let onode = graph.get(other); 
          onode.requirements.splice(onode.requirements.indexOf(node.id), 1); 
          if (onode.requirements.length === 0) { 
            graph.delete(other); 
            queue.push(onode); 
          }
        });
        a.work = undefined;
      });

      queue.sort((a, b) => a.id < b.id ? -1 : 1);
      available.forEach(worker => {
        let node = queue.shift();
        if (!node) return;
        worker.work = node;
        worker.time = 0;
        worker.end = sleep.get(node.id); 
      });

      if (!workers.some(worker => worker.work)) break;
      counter++;
    }
    console.log(counter);
  });
